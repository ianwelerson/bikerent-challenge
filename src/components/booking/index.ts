export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
export { default as BookingOverview } from './BookingOverview.vue'
export { default as BookingSuccess } from './BookingSuccess.vue'
