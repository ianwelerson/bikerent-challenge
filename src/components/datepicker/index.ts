export { default as Datepicker } from './DatepickerInput.vue'
