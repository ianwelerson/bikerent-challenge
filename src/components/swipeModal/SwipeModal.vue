<script lang="ts">
import { VueElement, defineComponent } from 'vue'

export default defineComponent({
  name: 'SwipeModal',
  props: {
    colorScheme: {
      type: String,
      default: 'white',
      validator: (value: string) => ['primary', 'white'].includes(value)
    },
    topSpacing: {
      type: Number,
      default: 120
    }
  },
  emits: ['close'],
  data: () => ({
    touchstart: 0,
    modalInititalPosition: 0,
    modalElement: null as VueElement | null,
    modalCloseAction: null as VueElement | null
  }),
  mounted() {
    this.modalElement = this.$refs['swipe-modal'] as VueElement
    this.modalCloseAction = this.$refs['swipe-close'] as VueElement

    this.modalCloseAction.addEventListener('touchstart', this.onTouchStart)
    this.modalCloseAction.addEventListener('touchend', this.onTouchEnd)

    this.modalElement.style.marginTop = `${this.topSpacing}px`
  },
  onBeforeUnmount() {
    this.modalCloseAction?.removeEventListener('touchstart', this.onTouchStart)
    this.modalCloseAction?.removeEventListener('touchend', this.onTouchEnd)
  },
  methods: {
    onTouchStart(event: TouchEvent) {
      this.touchstart = event.changedTouches[0].screenY
      this.modalInititalPosition = event.changedTouches[0].clientY
      this.modalCloseAction?.addEventListener('touchmove', this.onTouchMove)
    },
    onTouchEnd(event: TouchEvent) {
      if (this.modalElement) {
        this.modalElement.style.marginTop = `${this.topSpacing}px`
      }

      this.modalCloseAction?.removeEventListener('touchmove', this.onTouchMove)

      if (event.changedTouches[0].screenY - this.touchstart > 100) {
        this.closeModal()
      }
    },
    closeModal() {
      this.$emit('close')
    },
    onTouchMove(event: TouchEvent) {
      const distanceMoved = event.touches[0].clientY - this.modalInititalPosition
      const newMargin = distanceMoved > 0 ? this.topSpacing + distanceMoved : this.topSpacing // Ensure the movement is only to bottom
      if (this.modalElement) {
        this.modalElement.style.marginTop = `${newMargin}px`
      }
    }
  }
})
</script>

<template>
  <div :class="['swipe-modal', `swipe-modal--${colorScheme}`]">
    <div ref="swipe-modal" class="swipe-modal__wrapper">
      <div ref="swipe-close" class="swipe-modal__close"></div>
      <div class="swipe-modal__content">
        <slot />
      </div>
    </div>
  </div>
</template>

<style scoped lang="scss">
.swipe-modal {
  overscroll-behavior: none;
  position: fixed;
  left: 0;
  bottom: 0;
  height: 100vh;
  width: 100vw;
  z-index: 100;
  overflow: hidden;

  &::before {
    content: '';
    width: 100%;
    height: 100%;
    position: absolute;
    background: get-theme-color('black');
    opacity: 0.6;
    z-index: -1;
  }

  &__wrapper {
    height: 100%;
    width: 100%;
    border-radius: $border-radius-xl $border-radius-xl 0 0;
    display: flex;
    flex-direction: column;
  }

  &__close {
    padding: 12px 0 17px 0;
    width: 100%;
    align-self: center;
    position: relative;

    &::after {
      content: '';
      border-radius: 4px;
      width: 50px;
      height: 4px;
      position: absolute;
      left: 50%;
      transform: translate(-50%, 0);
      background: #d9d9d9; // Color missing in the varialble list
    }
  }

  &--primary {
    .swipe-modal__wrapper {
      background: get-theme-color('primary');
    }

    .swipe-modal__close {
      &::after {
        background: get-theme-color('white');
      }
    }
  }

  &--white {
    .swipe-modal__wrapper {
      background: get-theme-color('background');
    }

    .swipe-modal__close {
      &::after {
        background: #d9d9d9; // Color missing in the varialble list
      }
    }
  }

  &__content {
    overflow: auto;
    padding-bottom: 120px;
  }
}
</style>
