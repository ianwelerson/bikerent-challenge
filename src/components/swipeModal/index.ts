export { default as SwipeModal } from './SwipeModal.vue'
