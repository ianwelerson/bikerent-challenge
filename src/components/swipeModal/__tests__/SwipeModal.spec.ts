import { describe, expect, it } from 'vitest'

import { SwipeModal } from '@/components/swipeModal'
import { mount } from '@vue/test-utils'

describe('SwipeModal', () => {
  it('renders correctly', () => {
    const wrapper = mount(SwipeModal)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render the slot content', () => {
    const wrapper = mount(SwipeModal, {
      slots: {
        default: 'My Content'
      }
    })

    expect(wrapper.text()).equal('My Content')
  })
})
