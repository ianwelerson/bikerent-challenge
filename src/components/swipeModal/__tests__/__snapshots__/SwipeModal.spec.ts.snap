// Vitest Snapshot v1

exports[`SwipeModal > renders correctly 1`] = `
"<div class=\\"swipe-modal swipe-modal--white\\" data-v-424ad7ad=\\"\\">
  <div class=\\"swipe-modal__wrapper\\" data-v-424ad7ad=\\"\\" style=\\"margin-top: 120px;\\">
    <div class=\\"swipe-modal__close\\" data-v-424ad7ad=\\"\\"></div>
    <div class=\\"swipe-modal__content\\" data-v-424ad7ad=\\"\\"></div>
  </div>
</div>"
`;
